src/category_theory/essentially_small.lean

```diff
@@ -0,0 +1,210 @@
+we'll put a `category.{w}` instance on `shrink_homs C`.
+-/
+@[nolint has_inhabited_instance]
+def shrink_homs (C : Type u) := C
+
+namespace shrink_homs
+
+section
+variables {C' : Type*} -- a fresh variable with no category instance attached
+
+/-- Help the typechecker by explicitly translating from `C` to `shrink_homs C`. -/
+def to_shrink_homs {C' : Type*} (X : C') : shrink_homs C' := X
+/-- Help the typechecker by explicitly translating from `shrink_homs C` to `C`. -/
+def from_shrink_homs {C' : Type*} (X : shrink_homs C') : C' := X
+
+@[simp] lemma to_from (X : C') : from_shrink_homs (to_shrink_homs X) = X := rfl
+@[simp] lemma from_to (X : shrink_homs C') : to_shrink_homs (from_shrink_homs X) = X := rfl
+
+end
+
+variables (C) [locally_small.{w} C]
+
+@[simps]
+noncomputable
+instance : category.{w} (shrink_homs C) :=
+{ hom := λ X Y, shrink (from_shrink_homs X ⟶ from_shrink_homs Y),
+  id := λ X, equiv_shrink _ (𝟙 (from_shrink_homs X)),
+  comp := λ X Y Z f g,
+    equiv_shrink _ (((equiv_shrink _).symm f) ≫ ((equiv_shrink _).symm g)), }.
+
+/-- Implementation of `shrink_homs.equivalence`. -/
+@[simps]
+noncomputable
+def functor : C ⥤ shrink_homs C :=
+{ obj := λ X, to_shrink_homs X,
+  map := λ X Y f, equiv_shrink (X ⟶ Y) f, }
+
+/-- Implementation of `shrink_homs.equivalence`. -/
+@[simps]
+noncomputable
+def inverse : shrink_homs C ⥤ C :=
+{ obj := λ X, from_shrink_homs X,
+  map := λ X Y f, (equiv_shrink (from_shrink_homs X ⟶ from_shrink_homs Y)).symm f, }
+
+/--
+The categorical equivalence between `C` and `shrink_homs C`, when `C` is locally small.
+-/
+@[simps]
+noncomputable
+def equivalence : C ≌ shrink_homs C :=
+equivalence.mk (functor C) (inverse C)
+  (nat_iso.of_components (λ X, iso.refl X) (by tidy))
+  (nat_iso.of_components (λ X, iso.refl X) (by tidy))
+
+end shrink_homs
+
+/--
+A category is essentially small if and only if
+the underlying type of its skeleton (i.e. the "set" of isomorphism classes) is small,
+and it is locally small.
+-/
+theorem essentially_small_iff (C : Type u) [category.{v} C] :
+  essentially_small.{w} C ↔ small.{w} (skeleton C) ∧ locally_small.{w} C :=
+begin
+  -- This theorem is the only bit of real work in this file.
+  fsplit,
+  { intro h,
+    fsplit,
+    { rcases h with ⟨S, 𝒮, e , -⟩,
+      resetI,
+      refine ⟨⟨skeleton S, _, trivial⟩⟩,
+      exact e.skeleton_equiv, },
+    { resetI, apply_instance, }, },
+  { rintro ⟨⟨S, e, -⟩, L⟩,
+    resetI,
+    let e' := (shrink_homs.equivalence C).skeleton_equiv.symm,
+    refine ⟨⟨S, _, _, trivial⟩⟩,
+    apply induced_category.category (e'.trans e).symm,
+    refine (shrink_homs.equivalence C).trans
+      ((from_skeleton _).as_equivalence.symm.trans
+      ((induced_functor (e'.trans e).symm).as_equivalence.symm)), },
+end
+
+/--
+Any thin category is locally small.
+-/
+lemma locally_small_of_thin {C : Type u} [category.{v} C] (h : ∀ X Y : C, subsingleton (X ⟶ Y)) :
```

Comment:
Do you want `h` to be an inferred argument here? I think `subsingleton` is usually used with `[]`

Reply:
Done.